feature: adding `getmail` as an alternative to `fetchmail`

LucidityCrash opened the pull request:
# Description

Adding getmail as an alternative to Fetchmail

Fixes #2513

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [?] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [?] If necessary I have added tests that prove my fix is effective or that my feature works
- [?] New and existing unit tests pass locally with my changes

all of no_container.bats show fails 
test_helper.bats has a fails: 
 ✗ wait_for_smtp_port_in_container returns immediately when port found in 10400ms [10400]
318 tests, 9 failures, 4 skipped in 2025 seconds
I don't believe these are to do with my additions.

I've added a stub for testing but since getmail doesn't run as as daemon like fetchmail I'm not sure how to test.
Added some docs and basic config examples.

Not sure that putting the [options] section into a seperate file that is appended to each of the fetch configs as you might actually want different configs for different sources. 

Note I'm downloading the sid version of getmail6 as the stable one is quite a bit out of date.

casperklein commented on target/scripts/build/packages.sh:
```diff
@@ -97,6 +97,14 @@ function _install_packages
+function _install_getmail
+{
+  GETMAIL_SID_DEB_URL=http://http.us.debian.org/debian/pool/main/g/getmail6/getmail6_6.18.9-1_all.deb
+  curl -Lkso getmail6.deb  "${GETMAIL_SID_DEB_URL}" && \
+  dpkg -i getmail6.deb 2>&1 && \
```
```suggestion
  curl -Lkso getmail6.deb  "${GETMAIL_SID_DEB_URL}"
  dpkg -i getmail6.deb 2>&1
```

`&& \` are not needed, as we have `set -eE` at the top of the file.

LucidityCrash replied:
I've rebased against master again as I'd missed the latest ones with fail2ban in packages.sh and I've removed the && \ and made the variable "local" to match coding standard.

polarathene reviewed (changes requested):
Partial review as the browser crashed, but this was salvaged thankfully.

polarathene commented on target/scripts/build/packages.sh:
```diff
@@ -148,6 +148,14 @@ function _install_fail2ban
   rm fail2ban.deb fail2ban.deb.asc
 }
 
+function _install_getmail
+{
+  local GETMAIL_SID_DEB_URL='http://http.us.debian.org/debian/pool/main/g/getmail6/getmail6_6.18.9-1_all.deb'
+  curl -Lkso getmail6.deb  "${GETMAIL_SID_DEB_URL}"
+  dpkg -i getmail6.deb 2>&1
+  rm getmail6.deb
+}
```
Can you please justify the reason for this over just installing the [Bullseye package](https://packages.debian.org/bullseye/getmail6)?

LucidityCrash replied:
6.14 was released in Jan 2021, nearly 2 years ago.  The main reason for adding getmail as an alternative for fetchmail is the ability to utilise oauth, 6.18 contains fixes for depricated Google flows and fixes Microsoft oauth issues.  Seemed like a good idea to use it.

polarathene commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -1183,6 +1183,39 @@ EOF
+    mkdir "${GETMAILRC}"
+  fi
+
+  for FILE in ${CONFIGURATION}; do
+    if [[ -f ${FILE} ]]
+    then
+      USER=$(echo "${FILE}" | cut -d'-' -f 3| cut -d'.' -f1)
+      sed "s/PLACEHOLDER/${USER}/g" /etc/getmailrc_general > "${GETMAILRC}/getmailrc-${USER}.tmp"
+      cat "${GETMAILRC}/getmailrc-${USER}.tmp" "${FILE}" > "${GETMAILRC}/getmailrc-${USER}"
+      rm "${GETMAILRC}/getmailrc-${USER}.tmp"
+      GETMAIL_POLL_MINS=$((GETMAIL_POLL/60))
+      cat >"/etc/cron.d/getmail-${USER}" << EOF
+*/${GETMAIL_POLL_MINS} * * * * root /usr/bin/getmail --getmaildir /var/lib/getmail --rcfile "${GETMAILRC}/getmailrc-${USER}"
+EOF
+    else
+      cat /etc/getmailrc_general > "${GETMAILRC}/getmailrc"
+    fi
+  done
+  chmod -R 600 "${GETMAILRC}"
+  chown -R root:root "${GETMAILRC}"
+}
```
Same concern here with the `USER` as mentioned in an earlier review comment.

This answers the concern about `PLACEHOLDER` and default configs with the `sed` + `cat` lines.

Creating a separate cron job per user config seems like a :-1: from me. I think it'd be better to have a single cron job that can loop the command for each `--rcfile` arg? (_like you had in your `germail-wrapper.sh` script?_)

Is there a need for the `chmod` and `chown`? The `chmod` is fine, but `chown` should already be `root:root`?

polarathene replied:
Just to clarify, is `GETMAIL_POLL_MINS` the reason for citing a minimum of 60 seconds? That seems to be the case and perhaps related to your switch to using cron.

I would advise to instead, change the units from seconds to minutes. Simplifies the whole thing.

LucidityCrash replied:
Yep, will do, I went with seconds to keep it consistent with fetchmail.

polarathene commented on target/scripts/wrapper/getmail-wrapper.sh:
```diff
@@ -0,0 +1,10 @@
+#! /bin/bash
+while true; do
+    for file in /etc/getmailrc.d/getmailrc*; do
+        #/usr/bin/getmail --getmaildir /etc/getmailrc.d --rcfile getmailrc.file
+        if ! pgrep -f "${file}"$; then
+          /usr/bin/getmail --getmaildir /var/lib/getmail --rcfile "${file}"
+        fi
+    done
+    sleep 300
+done
```
`sleep 300` should be using `GETMAIL_POLL` ENV?

This file does not appear to be referenced anywhere in the PR, and supposedly replaced by the Cron job, is this file meant to be part of the PR?

As per the cronjob review comment, this could be repurposed to be part of a single cronjob, instead of one per user.

LucidityCrash replied:
yah you are correct, forgot to remove it.

LucidityCrash replied:
Repurposed as getmail-cron script.

polarathene commented on test/mail_getmail.bats:
```diff
@@ -0,0 +1,19 @@
+load 'test_helper/common'
+
+function setup_file() {
+  local PRIVATE_CONFIG
+  PRIVATE_CONFIG=$(duplicate_config_for_container .)
+
+  docker run -d --name mail_getmail \
+    -v "${PRIVATE_CONFIG}":/tmp/docker-mailserver \
+    -v "$(pwd)/test/test-files":/tmp/docker-mailserver-test:ro \
+    -e ENABLE_GETMAIL=1 \
+    --cap-add=NET_ADMIN \
+    -h mail.my-domain.com -t "${NAME}"
+
+  wait_for_finished_setup_in_container mail_getmail
+}
+
+function teardown_file() {
+  docker rm -f mail_getmail
+}
```
This test isn't serving any purpose, but we can discuss that once everything else from review is resolved.

I am aware that the current `getmail6` project on Github has their own tests that uses `docker-mailserver`, so we could probably reference what they're doing.

LucidityCrash replied:
Yeah ... not really sure what I can be doing here.  I'll take a look at the `getmail6 ` tests and see if I can update.

polarathene replied:
We'll discuss this once another maintainer reviews and this is all that is left :+1:

georglauterbach replied:
I adjusteded the test template, you will now only need to fill in the test functions. A good orientation are the Rspamd tests (as they are fairly recent) when it comes to using helper functions. See `test/tests/parallel/set1/spam_virus/rspamd...`.

polarathene commented on docs/content/config/environment.md:
```diff
@@ -445,6 +445,20 @@ Note: activate this only if you are confident in your bayes database for identif
+
+##### GETMAIL_POLL
+
+- **300** => `getmail` The number of seconds for the interval
```
```suggestion
- **300** => `getmail` The number of seconds for the interval (minimum 60)
```

LucidityCrash replied:
addressed ... changed GETMAIL_POLL to minutes and updated docs to give min and max values.

polarathene reviewed (changes requested):
Remainder of review completed :grinning:

LucidityCrash commented on target/bin/debug-getmail:
```diff
@@ -0,0 +1,11 @@
+# shellcheck source=../scripts/startup/setup-stack.sh
+source /usr/local/bin/setup-stack.sh
+
+_setup_getmail
+for file in /etc/getmailrc.d/getmailrc*; do
+    su -s /bin/sh -c /usr/bin/getmail --getmaildir /var/lib/getmail --rcfile "${file}" --dump
+done
```
Just following the fetchmail version ... like I said in the ticket about this I'm not that conversent with all of the things you are using so in a lot of cases I've used fetchmail as a template

polarathene replied:
I'll leave this up to @casperklein or @georglauterbach to decide if a change should be made here or resolve this feedback leaving it as-is :+1:

georglauterbach replied:
I do not see a reason to use `su` here as well. I'd remove it.

LucidityCrash commented on docs/content/config/advanced/mail-getmail.md:
````diff
@@ -0,0 +1,92 @@
+  - GETMAIL_POLL=60
+```
+
+You must specify a numeric argument which is a polling interval in seconds. The example above polls every minute for new mails.
````
Updated to use minutes, documented minimum and maximum values.

LucidityCrash commented on docs/content/config/advanced/mail-getmail.md:
````diff
@@ -0,0 +1,92 @@
+message_log = /var/log/mail/getmail-PLACEHOLDER.log
+```
+
+to override these mount your own common settings to /etc/getmailrc_general
````
I've removed the need for the message_log  setting in this file and tried to clarify that mounting a file replaces this one so all options you want need to be defined.

polarathene commented on docs/content/config/advanced/mail-getmail.md:
````diff
@@ -0,0 +1,88 @@
+    arguments =("-d","user1@example.com")
+    ```
+
+### POP3 Configuration 
+
+!!! example
````
```suggestion
### POP3 Configuration

Just like the IMAP example above, but instead via POP3 protocol if you prefer that over IMAP.

!!! example
```

polarathene commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,88 @@
+    [retriever]
+    type = SimplePOP3Retriever
+    server = pop3.gmail.com
+    username = username
+    password = secret
```
```suggestion
    username = alice
    password = notsecure
```

polarathene commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,88 @@
+    type = MDA_external
+    path = /usr/lib/dovecot/deliver
+    allow_root_commands = true
+    arguments =("-d","user2@example.com")
```
Just keeping it consistent with the IMAP config example :)

```suggestion
    arguments =("-d","user1@example.com")
```

polarathene commented on docs/content/config/environment.md:
```diff
@@ -445,6 +445,20 @@ Note: activate this only if you are confident in your bayes database for identif
+
+##### GETMAIL_POLL
+
+- **5** => `getmail` The number of minutes for the interval (minimum 1, maximum 60)
```
```suggestion
- **5** => `getmail` The number of minutes for the interval (min: 1, max: 60)
```

polarathene reviewed (changes requested):
I used [this online crontab tool](https://crontab.guru/#*/29_*_*_*_*), which should link to an example of every 29th minute. You'll see that it's only within an hour, then it resets. So every 50th minute would be always at 50 minutes past the current hour, technically every hour at that point.. same with every 31 minutes.

Thus I think the actual max is 30? (Every 30 minutes) Otherwise it's effectively hourly from then on, and anything after 60 likewise ends up being every hour (just with no offset).

I don't know `cron` that well, the other maintainers can provide feedback on that.

---

I'm going to batch commit the suggestions, so only the remaining feedback after (3 items left) that needs to be addressed :+1: 

- [`su -s` decision deferred to another reviewer](https://github.com/docker-mailserver/docker-mailserver/pull/2803#discussion_r988384985).
- [`xoauth` docs link](https://github.com/docker-mailserver/docker-mailserver/pull/2803#discussion_r990998952).
- [Adding tests](https://github.com/docker-mailserver/docker-mailserver/pull/2803#discussion_r988395196) (_once everything else is approved_).

georglauterbach commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -1183,6 +1183,44 @@ EOF
+{
+  _log 'trace' 'Preparing Getmail configuration'
+
+  local CONFIGURATION GETMAILRC ID CONFIGS
+
+  CONFIGURATION='/tmp/docker-mailserver/getmail-*.cf'
```
We do not need `CONFIGURATION` to be in a variable really, just use idiomatic globbing for the loop:

```suggestion
  local GETMAILRC ID CONFIGS

```

georglauterbach commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -1183,6 +1183,44 @@ EOF
+
+  # Generate getmailrc configs, starting with the `/etc/getmailrc_general` base config,
+  # Add a unique `message_log` config, then append users own config to the end.
+  for FILE in ${CONFIGURATION}; do
```
```suggestion
  for FILE in /tmp/docker-mailserver/getmail-*.cf
```

This is actually idiomatic in Bash. Iterating over a variable that just happens to expand into a string with whitespaces is not, especially because it will not work when you start properly quoting it :)

LucidityCrash replied:
Makes sense to me.

georglauterbach commented on target/bin/getmail-cron:
```diff
@@ -0,0 +1,6 @@
+#! /bin/bash
+for file in /etc/getmailrc.d/getmailrc*; do
+  if ! pgrep -f "${file}"$ > /dev/null 2>&1; then
+    /usr/bin/getmail --getmaildir /var/lib/getmail --rcfile "${file}"
+  fi
+done
```
```suggestion
#! /bin/bash

for FILE in /etc/getmailrc.d/getmailrc*
do
  if ! pgrep -f "${FILE}"$ > /dev/null 2>&1
  then
    /usr/bin/getmail --getmaildir /var/lib/getmail --rcfile "${FILE}"
  fi
done
```

Is the `$` after `"${FILE}"` intended?

LucidityCrash replied:
it was intended, to make sure it was matching at the end of the command.  Not sure why I did that though  ... could probably take that out

georglauterbach reviewed (changes requested):
Looks very good, just some minor stylistic issues. I hope you can just batch-commit my suggestions :)

georglauterbach commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -1183,6 +1183,44 @@ EOF
+    if [[ -f ${FILE} ]]
+    then
+      CONFIGS=true
+      ID=$(echo "${FILE}" | cut -d'-' -f 3| cut -d'.' -f1)
```
```suggestion
      ID=$(cut -d '-' -f 3 <<< "${FILE}" | cut -d '.' -f 1)
```

georglauterbach commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -1183,6 +1183,44 @@ EOF
+      local GETMAIL_CONFIG="${GETMAILRC}/getmailrc-${ID}"
+      cat /etc/getmailrc_general > "${GETMAIL_CONFIG}.tmp"
+      echo -e "message_log = /var/log/mail/getmail-${ID}.log\n" >> "${GETMAIL_CONFIG}.tmp"
+      cat "${GETMAIL_CONFIG}.tmp" "${FILE}" > "${GETMAIL_CONFIG}"
```
```suggestion
      cat "${GETMAIL_CONFIG}.tmp" "${FILE}" >"${GETMAIL_CONFIG}"
```

LucidityCrash commented:
> Thus I think the actual max is 30? (Every 30 minutes) Otherwise it's effectively hourly from then on, and anything after 60 likewise ends up being every hour (just with no offset).

Yeah you are correct ... nothing above 30 really makes any sense.

LucidityCrash commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,88 @@
+
+[getmail-website]: https://www.getmail.org
+[getmail-docs]: https://getmail6.org/configuration.html
+[getmail-gmail-xoauth]: https://www.bytereef.org/howto/oauth2/getmail.html
```
yep you are correct, nothing above 30 really makes any sense.
Updated docs to add pointers to the xoauth example configs in the getmail6 github repo.

georglauterbach commented:
This can now be merged when the comments are resolved :)

georglauterbach commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -1183,6 +1183,43 @@ EOF
+
+  # Generate getmailrc configs, starting with the `/etc/getmailrc_general` base config,
+  # Add a unique `message_log` config, then append users own config to the end.
+  for FILE in /tmp/docker-mailserver/getmail-*.cf
```
```suggestion
  for FILE in /tmp/docker-mailserver/getmail-*.cf
  do
```

georglauterbach commented:
Applied my own review feedback as well to make the final reviews faster :)

georglauterbach reviewed (dismissed):
From a scripts perspective, LGTM. I hope all tests pass now :)

polarathene commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+
+[getmail-website]: https://www.getmail.org
+[getmail-docs]: https://getmail6.org/configuration.html
+[getmail-gmail-xoauth-12]: https://github.com/getmail6/getmail6/blob/master/docs/getmailrc-examples#L286
```
```suggestion
[getmail-docs-xoauth-12]: https://github.com/getmail6/getmail6/blob/1f95606156231f1e074ba62a9baa64f892a92ef8/docs/getmailrc-examples#L286
```

polarathene commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+[getmail-website]: https://www.getmail.org
+[getmail-docs]: https://getmail6.org/configuration.html
+[getmail-gmail-xoauth-12]: https://github.com/getmail6/getmail6/blob/master/docs/getmailrc-examples#L286
+[getmail-gmail-xoauth-12]: https://github.com/getmail6/getmail6/blob/master/docs/getmailrc-examples#L351
```
```suggestion
[getmail-docs-xoauth-13]: https://github.com/getmail6/getmail6/blob/1f95606156231f1e074ba62a9baa64f892a92ef8/docs/getmailrc-examples#L351
```

polarathene commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+
+### XOAUTH2 Authentication
+
+It is possible to utilize the getmail-gmail-xoauth-token helper to provide authentication using xoauth2 for gmail [example 12][getmail-gmail-xoauth-12] or Microsoft Office 365 [example 13][getmail-gmail-xoauth-12]
```
```suggestion
It is possible to utilize the `getmail-gmail-xoauth-tokens` helper to provide authentication using `xoauth2` for [gmail (example 12)][getmail-docs-xoauth-12] or [Microsoft Office 365 (example 13)][getmail-docs-xoauth-13]
```

polarathene reviewed (changes requested):
I will batch commit these changes.

Fixes some typos, and uses permalinks instead of referencing lines on the master branch which is not a stable reference we can rely on.

polarathene reviewed (dismissed):
We still may want to consider tests.

If I understand this feature correctly. We could test it by having two DMS containers running, and treating one of them as the remote mail-server (_the role of gmail for example_)?

Additionally, [here is the `getmail6` bats test file](https://github.com/getmail6/getmail6/blob/1f95606156231f1e074ba62a9baa64f892a92ef8/test/test_getmail_with_docker_mailserver.bats) that uses DMS to handle it's testing. I haven't looked through it properly yet.

georglauterbach commented:
Yes, I'd like to see some tests as well - good idea :)

polarathene commented:
I've assigned myself, but welcome anyone else to tackle the remaining task of getting some tests sorted.

This is a low priority for me and may not be something I can work on until December.

polarathene commented:
There is an issue with building the image during `packages.sh`. I'm going to take a guess that the URL for the package isn't valid? If so that may be an issue and we'd need something a bit more reliable / stable :sweat_smile: 

```
#18 82.52 dpkg-deb: error: 'getmail6.deb' is not a Debian format archive
#18 82.52 dpkg: error processing archive getmail6.deb (--install):
#18 82.52  dpkg-deb --control subprocess returned error exit status 2
#18 82.52 Errors were encountered while processing:
#18 82.52  getmail6.deb
```

This PR will be otherwise addressed after the rest of the test suite is revised. As the current fetchmail tests aren't that extensive, I'm also open towards relaxing the requirement on this alternative to just have parity with the fetchmail tests. Although proper testing would be better, it'll be easier to do that at a later date.

If the package install can't be more reliable like dovecot, rspamd and fail2ban are, perhaps we can delay this until the next base image upgrade, but that could be mid-year.

LucidityCrash commented:
Yeah I just rebased and realised that sid removes the older versions of
packages, so I updated the url but as you say that won't work long term ...
perhaps picking up the package from an Ubuntu mirror ... they don't seem to
remove the older packages.

On Fri, 27 Jan 2023, 22:49 Brennan Kinney, ***@***.***> wrote:

> There is an issue with building the image during packages.sh. I'm going
> to take a guess that the URL for the package isn't valid? If so that may be
> an issue and we'd need something a bit more reliable / stable 😅
>
> #18 82.52 dpkg-deb: error: 'getmail6.deb' is not a Debian format archive
>
> #18 82.52 dpkg: error processing archive getmail6.deb (--install):
>
> #18 82.52  dpkg-deb --control subprocess returned error exit status 2
>
> #18 82.52 Errors were encountered while processing:
>
> #18 82.52  getmail6.deb
>
>
> This PR will be otherwise addressed after the rest of the test suite is
> revised. As the current fetchmail tests aren't that extensive, I'm also
> open towards relaxing the requirement on this alternative to just have
> parity with the fetchmail tests. Although proper testing would be better,
> it'll be easier to do that at a later date.
>
> If the package install can't be more reliable like dovecot, rspamd and
> fail2ban are, perhaps we can delay this until the next base image upgrade,
> but that could be mid-year.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docker-mailserver/docker-mailserver/pull/2803#issuecomment-1407163271>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALWZJK4ZW4HW6AOWX7ZLQO3WURGHLANCNFSM6AAAAAAQY3HJXQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

polarathene commented:
> perhaps picking up the package from an Ubuntu mirror ... they don't seem to remove the older packages.

I just noticed that this is only python, so we can probably build it directly from the Github source / release :sweat_smile: 

- [`pip install getmail6`](https://github.com/getmail6/getmail6#installation). Not sure how other maintainers feel about that approach though.
- Debian package does have the [6.18 release available in their repo](https://salsa.debian.org/python-team/packages/getmail6), so it'll be part of the new release this year :+1:

georglauterbach commented:
> I just noticed that this is only python, so we can probably build it directly from the Github source / release sweat_smile
> 
>     * [`pip install getmail6`](https://github.com/getmail6/getmail6#installation). Not sure how other maintainers feel about that approach though.

Please don't. I would like to have a little Python in this project as possible..

polarathene commented:
> Please don't. I would like to have a little Python in this project as possible..

Why did the original issue and this PR get as far as it has then? The project is python based, so to approve it will allow more python.

I can understand not wanting to use `pip`, I had thought that was already available in the image, but adding `python3-pip` seems like it'd add 255MB size to the image :sweat_smile: 

Debian seems to have an [alternative build system](https://salsa.debian.org/python-team/packages/getmail6/-/blob/debian/master/debian/rules) rather than via `pip` for it, and I guess removes all the build-time only dependencies once packaged. Do we use the ubuntu mirror mentioned, or wait until next Debian release later this year?

georglauterbach commented:
> > Please don't. I would like to have a little Python in this project as possible..
> 
> Why did the original issue and this PR get as far as it has then? The project is python based, so to approve it will allow more python.

I should have been more precise, sorry. I am not entirely against Python in this project (although, I think that would be nice; not feasible ATM though). I would just like to maintain _as little Python code as possible_. This issue/PR does not actually introduce (more) Python code we need to maintain to DMS, so it should be fine, even though I am not pumped about getmail being written in Python 🤣

> I can understand not wanting to use `pip`, I had thought that was already available in the image, but adding `python3-pip` seems like it'd add 255MB size to the image sweat_smile

Yeah, thats definitely an exclusion criteria 😆
 
> Debian seems to have an [alternative build system](https://salsa.debian.org/python-team/packages/getmail6/-/blob/debian/master/debian/rules) rather than via `pip` for it, and I guess removes all the build-time only dependencies once packaged. Do we use the ubuntu mirror mentioned, or wait until next Debian release later this year?

👍🏼

fdemmer commented:
> adding python3-pip seems like it'd add 255MB size to the image

No idea how you managed that... maybe that includes the whole Python interpreter with its standardlib again? pip itself is much smaller and getmail6 is just a few hundred kb.

I just built current master and a modified `packages.sh` from this pr's branch and those are the image sizes:
- master: 613MB
- getmail-pr: 622MB

this is the `_install_getmail` function I used:

```
function _install_getmail
{
  local GETMAIL_PYPI_URL='https://files.pythonhosted.org/packages/31/f7/9e3d0204b4b5dd330c572807bd2021436ebb261be206b87b7d17e3259a78/getmail6-6.18.12-py3-none-any.whl'

  _log 'debug' 'Installing getmail6'
  apt-get "${QUIET}" --no-install-recommends install python3-pip
  pip3 install ${GETMAIL_PYPI_URL}
  ln -s /usr/local/bin/getmail /usr/bin/getmail
}
```

By using the exact URL with hash we are guaranteed to get the same source package on every build. There is no Python code to maintain and it is easy to follow upstream getmail6 updates (or not).

polarathene commented:
> No idea how you managed that... maybe that includes the whole Python interpreter with its standardlib again? pip itself is much smaller and getmail6 is just a few hundred kb.

My bad! I forgot to use `--no-install-recommends` :sweat_smile: 

With that added it adds 6.4MB to image weight which seems ok.

> `local GETMAIL_PYPI_URL='https://files.pythonhosted.org/packages/31/f7/9e3d0204b4b5dd330c572807bd2021436ebb261be206b87b7d17e3259a78/getmail6-6.18.12-py3-none-any.whl'`

I don't really like that. Any reason we can't use something like:

- `pip install 'getmail6>=6.18.12'` (_not familiar with this package manager and what range that permits vs semver patch_)
- `pip install 'getmail6==6.18.12'` (_version pinned_)

That would install the [`getmail6` package from PyPi](https://pypi.org/project/getmail6/).

fdemmer commented:
Sure, my reason for using the URL, was that it contains a hash of the exact file in addition to the version in the filename for a repeatable build. `files.pythonhosted.org` is the server PyPI uses.

Alternatively it is of course possible to just [pin the version or a range](https://peps.python.org/pep-0440/#version-specifiers).

- `pip install 'getmail6>=6.18.12'` installs 6.18.12 or "7.8.9" or any version higher
- `pip install 'getmail6~=6.18.12'` installs 6.18.x but not 8.13 etc.

Depends on that the requirements are ;) ... reproducable build should pin the version and use a hash to defend against compromised PyPI or someone switching the package index `pip` uses during the build process.

more documentation:
- https://pip.pypa.io/en/stable/topics/repeatable-installs/
- https://pip.pypa.io/en/stable/topics/configuration/#environment-variables

polarathene commented:
> If all that is not required, `pip install 'getmail6~=6.18.0'` is probably best to get 6.18 and any patch releases on 6.18 level.

That sounds perfect for us :)

I understand the value of reproducible builds, but we don't have that with our images as other packages aren't pinned.

georglauterbach commented:
Moved to v12.1.0.

georglauterbach commented:
I'll be tackling this after v12.1.0 and my current:upcoming PRs are done.

georglauterbach commented:
@LucidityCrash thare was a significant amount of changes for DMS lately, which means I'd require you to please rebase this PR and resolve the merge conflicts. I am willing to add this to DMS, but looking at #3289, I have too much to do to tackle this (especially because it's a new feature).

If you lack the time, please tell. I would then like to close this PR (see <https://github.com/docker-mailserver/docker-mailserver/issues/3289#issuecomment-1520172545> on why). I know that @polarathene is more than packed with work, and he won't realistically find the time to tackle this PR either.

I think your changes are actually quite good, it's just that this PR has stalled. Please reach out back to me :)

LucidityCrash commented:
I'll endeavour to get a rebase done this week, is that acceptable?

On Mon, 24 Apr 2023, 14:45 Georg Lauterbach, ***@***.***>
wrote:

> @LucidityCrash <https://github.com/LucidityCrash> thare was a significant
> amount of changes for DMS lately, which means I'd require you to please
> rebase this PR and resolve the merge conflicts. I am willing to add this to
> DMS, but looking at #3289
> <https://github.com/docker-mailserver/docker-mailserver/issues/3289>, I
> have too much to do to tackle this (especially because it's a new feature).
>
> If you lack the time, please tell. I would then like to close this PR (see #3289
> (comment)
> <https://github.com/docker-mailserver/docker-mailserver/issues/3289#issuecomment-1520172545>
> on why). I know that @polarathene <https://github.com/polarathene> is
> more than packed with work, and he won't realistically find the time to
> tackle this PR either.
>
> I think your changes are actually quite good, it's just that this PR has
> stalled. Please reach out back to me :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docker-mailserver/docker-mailserver/pull/2803#issuecomment-1520190820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALWZJK3ZCMQ5G5JYADWZ7OLXCZ7WNANCNFSM6AAAAAAQY3HJXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

georglauterbach commented:
> I'll endeavour to get a rebase done this week, is that acceptable?

Absolutely :) I think I can assist you with tests thereafter. This PR is marked for v13.0.0, and v12.1.0 preceedes v13.0.0. v12.1.0 has no release PR yet, so there is time.

polarathene commented:
**EDIT:** Forgot to post this earlier sorry :sweat_smile: 

> he won't realistically find the time to tackle this PR either.

I cannot commit to finishing this sorry.

It was discussed that the blocker of the test here is a non-issue. It could just copy the [`fetchmail` test](https://github.com/docker-mailserver/docker-mailserver/blob/223c76632013aa3e9207279a6c13fe5c3ffd077e/test/tests/parallel/set1/fetchmail.bats) and adapt to that (_can ignore the parallel tests_). That test effectively just checks config is working, rather than the actual functionality of the feature. Probably should also add `getmail` process [into this test](https://github.com/docker-mailserver/docker-mailserver/blob/223c76632013aa3e9207279a6c13fe5c3ffd077e/test/tests/parallel/set3/container_configuration/process_check_restart.bats#L46) too.

Since the equivalent `fetchmail` feature does not have better tests and is kept around, I have no issue with `getmail` having that same treatment.

Other than the basic tests (_mostly copy/paste with few changes_) - as per earlier discussion in the PR above, [`getmail` should be installed via `pip`](https://github.com/docker-mailserver/docker-mailserver/pull/2803#issuecomment-1420313402): `pip install 'getmail6~=6.18.0'`

georglauterbach commented:
v12.1.0 is expected to be released on May 10. This PR can be merged afterwards - please rebase until then.

LucidityCrash commented:
Moved the install of getmail to pip as per the comments.

georglauterbach commented:
Very nice! I will assist you with the tests. I'll update you when I had a look.

georglauterbach commented:
Checks seem to pass 🎉

georglauterbach commented:
I will provide a review for this PR today. When all of the feedback is resolved, we can have a look at tests.

LucidityCrash commented:
I upgraded a couple of days ago.  Just did both OS and ad guard at the same
time ... only thing you need to do is go to the AdGuardHome config page in
OpnSense Services and tick the new Primary DNS option, then restart AdGuard
and then DHCP services.

On Sat, 6 May 2023, 11:44 Georg Lauterbach, ***@***.***>
wrote:

> I will provide a review for this PR today. When all of the feedback is
> resolved, we can have a look at tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docker-mailserver/docker-mailserver/pull/2803#issuecomment-1537114094>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALWZJKZHYO3NWYISYGV43FTXEYTSBANCNFSM6AAAAAAQY3HJXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

georglauterbach commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+title: 'Advanced | Email Gathering with Getmail'
+---
+
+To enable the [getmail][getmail-website] service to retrieve e-mails set the environment variable `ENABLE_GETMAIL` to `1`. Your `docker-compose.yml` file should include the following:
```
```suggestion
To enable the [getmail][getmail-website] service to retrieve e-mails set the environment variable `ENABLE_GETMAIL` to `1`. Your `compose.yaml` file should include the following:
```

georglauterbach commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+
+In your DMS config volume (eg: `docker-data/dms/config/`), create a `getmail-<ID>.cf` file for each remote account that you want to retrieve mail and store into a local DMS account. `<ID>` should be replaced by you, and is just the rest of the filename (eg: `getmail-example.cf`). The contents of each file should be configuration like documented below.
+
+Your `docker-mailserver` folder should look similar to this example:
```
```suggestion
The directory structure should similar to this:
```

georglauterbach commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+
+### Common Options
+
+The default options added to each `getmail` config are:
```
```suggestion
The default options added to each `getmail` config are:

```

georglauterbach commented on docs/content/config/advanced/mail-getmail.md:
````diff
@@ -0,0 +1,103 @@
+
+If you want to use a different base config, mount a file to `/etc/getmailrc_general`. This file will replace the default "Common Options" base config above, that all `getmail-<ID>.cf` files will extend with their configs when used.
+
+### IMAP Configuration 
+
+This example will:
+
+1. Connect to the remote IMAP server from Gmail.
+2. Retrieve mail from the gmail account `alice` with password `notsecure`.
+3. Store any mail retrieved from the remote mail-server into DMS for the `user1@example.com` account that DMS manages.
+
+!!! example
+    ```getmailrc
+    [retriever]
+    type = SimpleIMAPRetriever
+    server = imap.gmail.com
+    username = alice
+    password = notsecure
+
+    [destination]
+    type = MDA_external
+    path = /usr/lib/dovecot/deliver
+    allow_root_commands = true
+    arguments =("-d","user1@example.com")
+    ```
````
```suggestion
??? example "IMAP Configuration" 

    This example will:

    1. Connect to the remote IMAP server from Gmail.
    2. Retrieve mail from the gmail account `alice` with password `notsecure`.
    3. Store any mail retrieved from the remote mail-server into DMS for the `user1@example.com` account that DMS manages.

    ```getmailrc
    [retriever]
    type = SimpleIMAPRetriever
    server = imap.gmail.com
    username = alice
    password = notsecure
    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","user1@example.com")
    ```
```

georglauterbach commented on docs/content/config/advanced/mail-getmail.md:
````diff
@@ -0,0 +1,103 @@
+    arguments =("-d","user1@example.com")
+    ```
+
+### POP3 Configuration
+
+Just like the IMAP example above, but instead via POP3 protocol if you prefer that over IMAP.
+
+!!! example
+    ```getmailrc
+    [retriever]
+    type = SimplePOP3Retriever
+    server = pop3.gmail.com
+    username = alice
+    password = notsecure
+
+    [destination]
+    type = MDA_external
+    path = /usr/lib/dovecot/deliver
+    allow_root_commands = true
+    arguments =("-d","user1@example.com")
+    ```
````
```suggestion
??? example "POP3 Configuration"

    Just like the IMAP example above, but instead via POP3 protocol if you prefer that over IMAP.

    ```getmailrc
    [retriever]
    type = SimplePOP3Retriever
    server = pop3.gmail.com
    username = alice
    password = notsecure
    [destination]
    type = MDA_external
    path = /usr/lib/dovecot/deliver
    allow_root_commands = true
    arguments =("-d","user1@example.com")
    ```
```

georglauterbach commented on docs/content/config/advanced/mail-getmail.md:
```diff
@@ -0,0 +1,103 @@
+
+### Polling Interval
+
+By default the `getmail` service checks external mail accounts for new mail every 5 minutes. That polling interval is configurable via the `GETMAIL_POLL` ENV variable, with a value in minutes (_default: 5, min: 1, max: 30_):
```
```suggestion
By default the `getmail` service checks external mail accounts for new mail every 5 minutes. That polling interval is configurable via the `GETMAIL_POLL` ENV variable, with a value in minutes (_default: 5, min: 1, max: 30_):

```

georglauterbach commented on docs/content/config/environment.md:
```diff
@@ -546,6 +546,20 @@ Note: activate this only if you are confident in your bayes database for identif
+
+#### Getmail
+
+##### ENABLE_GETMAIL
+
+- **0** => `getmail` disabled
+- 1 => `getmail` enabled
+
+##### GETMAIL_POLL
+
+- **5** => `getmail` The number of minutes for the interval (min: 1, max: 30)
+
+Note: Fetchmail and Getmail are mutually exclusive
```
```suggestion
##### ENABLE_GETMAIL

Enable or disable `getmail`. Note: Fetchmail ([`ENABLE_FETCHMAIL`](#enable_fetchmail)) and Getmail are mutually exclusive!

- **0** => Disabled
- 1 => Enabled

##### GETMAIL_POLL

The number of minutes for the interval. Min: 1; Max: 30; Default: 5.
```

georglauterbach commented on mailserver.env:
```diff
@@ -382,6 +382,15 @@ ENABLE_FETCHMAIL=0
 # The interval to fetch mail in seconds
 FETCHMAIL_POLL=300
 
+# -----------------------------------------------
+# --- Getmail Section -------------------------
+# -----------------------------------------------
+
+ENABLE_GETMAIL=0
+
+# The interval duration in minutes between mail fetches (min: 1, max: 30)
+GETMAIL_POLL=5
+
```
```suggestion
# Enable or disable `getmail`.
# Note: Fetchmail (ENABLE_FETCHMAIL) and Getmail are mutually exclusive!
#
# - **0** => Disabled
# - 1 => Enabled
ENABLE_GETMAIL=0

# The number of minutes for the interval. Min: 1; Max: 30.
GETMAIL_POLL=5
```

georglauterbach commented on target/scripts/build/packages.sh:
```diff
@@ -201,6 +201,18 @@ function _install_fail2ban
+  _log 'debug' 'Installing getmail6'
+  apt-get "${QUIET}" --no-install-recommends install python3-pip
+  pip3 install 'getmail6~=6.18.12'
+  ln -s /usr/local/bin/getmail /usr/bin/getmail
+  ln -s /usr/local/bin/getmail-gmail-xoauth-tokens /usr/bin/getmail-gmail-xoauth-tokens
```
AFAIK `/usr/bin/` is already in `${PATH}`.

```suggestion
  apt-get "${QUIET}" purge python3-pip
  apt-get "${QUIET}" autoremove
```

LucidityCrash replied:
those are linking from /usr/local/bin into /usr/bin ... you say /usr/bin is in the path but is /usr/local/bin ... Based on my testing I don't believe so.
I'll admit not having tested the requirement to link getmail into /usr/bin but if you don't link getmail-gmail-xauth-tokens then the getmail rc file has to specify the full path in the password_command option.

georglauterbach replied:
```console
root@mail:/# echo "${PATH}"
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

There really is no difference and we do not need the `ln`, I'm fairly certain.

LucidityCrash replied:
Well the reason I put the line
` ln -s /usr/local/bin/getmail-gmail-xoauth-tokens /usr/bin/getmail-gmail-xoauth-tokens`
was because without it I got errors with this in my config : 
`password_command = ("getmail-gmail-xoauth-tokens", "/var/lib/getmail/configfile.json")`
the error I got was : 
`Configuration error: Program "getmail-gmail-xoauth-tokens" not found`

and the reason the /usr/bin/getmail link is there is because the cron job calls it from there:
`/usr/bin/getmail --getmaildir /var/lib/getmail --rcfile "${FILE}"`

Ultimately it is the difference between the deb install which put the files in /usr/bin/ and the  pip install which put them in /usr/local/bin/

If you really don't want the links then getmail-cron and debug-getmail will need to be updated and we should put a note in the docs that examples given in the links to getmail6's docks wont work as is and people will have to put in the full path to the getmail-gmail-xoauth-tokens helper.

georglauterbach replied:
I see; then we should change to `/usr/bin/` in our scripts, and re-introduce the links as well though to make us easy. Sorry for the noise then; will you take care of this or shall I do it?

georglauterbach replied:
I re-applied the `ln` commands in [5b19dbc](https://github.com/docker-mailserver/docker-mailserver/pull/2803/commits/5b19dbc91ff702eb8beced5d5eeb464209d9e7fe)

georglauterbach commented on target/scripts/start-mailserver.sh:
```diff
@@ -97,6 +98,11 @@ function _register_functions
   # needs to come after _setup_postfix_early
   _register_setup_function '_setup_spoof_protection'
 
+  if [[ ${ENABLE_GETMAIL} -eq 1 ]]
+  then
+    _register_setup_function '_setup_getmail'
+  fi
+
```
```suggestion
[[ ${ENABLE_GETMAIL} -eq 1 ]] && _register_setup_function '_setup_getmail'
```

georglauterbach commented on target/scripts/startup/setup.d/getmail.sh:
```diff
@@ -0,0 +1,39 @@
+  GETMAILRC='/etc/getmailrc.d'
+  CONFIGS=false
+
+  if [[ ! -d ${GETMAILRC} ]]
+  then
+    mkdir "${GETMAILRC}"
+  fi
```
```suggestion
  mkdir -p "${GETMAILRC}"
```

georglauterbach commented on target/scripts/startup/setup.d/getmail.sh:
```diff
@@ -0,0 +1,39 @@
+  local GETMAILRC ID CONFIGS
+
+  GETMAILRC='/etc/getmailrc.d'
+  CONFIGS=false
```
While I started out on this project using `false` and `true` as well, it has become the norm in DMS scripts to use `0` for `false` and `1` for `true`. I you want to know about why, you can ask me of course :)

```suggestion
  CONFIGS=0
```

georglauterbach commented on target/scripts/startup/setup.d/getmail.sh:
```diff
@@ -0,0 +1,39 @@
+  do
+    if [[ -f ${FILE} ]]
+    then
+      CONFIGS=true
```
```suggestion
      CONFIGS=1
```

georglauterbach commented on target/scripts/startup/setup.d/getmail.sh:
```diff
@@ -0,0 +1,39 @@
+      cat "${GETMAIL_CONFIG}.tmp" "${FILE}" >"${GETMAIL_CONFIG}"
+      rm "${GETMAIL_CONFIG}.tmp"
+    fi
+  done
+  if [[ ${CONFIGS} == true ]]
```
```suggestion
  done

  if [[ ${CONFIGS} -eq 1 ]]
```

georglauterbach commented on target/scripts/startup/setup.d/getmail.sh:
```diff
@@ -0,0 +1,39 @@
+  done
+  if [[ ${CONFIGS} == true ]]
+  then
+    cat >"/etc/cron.d/getmail" << EOF
```
```suggestion
    cat >/etc/cron.d/getmail << EOF
```

georglauterbach reviewed (changes requested):
Minor changes; I will commit them immediately to not waste time :)

georglauterbach commented on target/scripts/start-mailserver.sh:
```diff
@@ -97,6 +98,7 @@ function _register_functions
   # needs to come after _setup_postfix_early
   _register_setup_function '_setup_spoof_protection'
 
+[[ ${ENABLE_GETMAIL} -eq 1 ]] && _register_setup_function '_setup_getmail'
```
```suggestion
  [[ ${ENABLE_GETMAIL} -eq 1 ]] && _register_setup_function '_setup_getmail'

```

georglauterbach commented on mailserver.env:
```diff
@@ -382,6 +382,15 @@ ENABLE_FETCHMAIL=0
+ENABLE_GETMAIL=0
+
+# The number of minutes for the interval. Min: 1; Max: 30.
+GETMAIL_POLL=5
```
```suggestion
GETMAIL_POLL=5

```

georglauterbach commented:
Review done; all looks good now. I will see what we can do for tests.

georglauterbach commented:
### Things I've Identified That We Can Test

1. When `ENABLE_GETMAIL` is set, whether all default config files exists (e.g., `getmailrc`, etc.), and whether they contain the correct content
2. When ´GETMAIL_POLL` is set, whether it's adjustments work
3. We should be able to test whether `debug-getmail` does as it should

@LucidityCrash These are the basic tests I'd expect we should have when mergeing this PR. I think the already present test files should provide a good orientation when writing tests for getmail. I really do not expect the tests for getmail to be very elaborate, but they should cover the absolute basics. Just reach out to me when you have any questions.

georglauterbach commented on target/scripts/build/packages.sh:
```diff
@@ -201,6 +201,18 @@ function _install_fail2ban
+  _log 'debug' 'Installing getmail6'
+  apt-get "${QUIET}" --no-install-recommends install python3-pip
+  pip3 install 'getmail6~=6.18.12'
+  apt-get "${QUIET}" purge python3-pip
```
```suggestion
  ln -s /usr/local/bin/getmail /usr/bin/getmail
  ln -s /usr/local/bin/getmail-gmail-xoauth-tokens /usr/bin/getmail-gmail-xoauth-tokens
  apt-get "${QUIET}" purge python3-pip
```

georglauterbach commented:
Ah, and I forgot: you need to move the test file to `test/tests/parallel/set1/`.

georglauterbach commented:
v12.1.0 just dropped. When tests are added to this PR, we can go ahead and merge it 🚀

georglauterbach commented:
I provided the basic tests, but cannot guarantee that the getmail setup is now functioning properly. For what it's worth, the changes look good. We could have better getmail coverage, but I guess users of getmail will complain if there's something that's not rights. by now, I just want this to get merged, and to me, it actually looks quite decent.

I'll wait for @polarathene's opinion, but he's low on time, so if you want to skip the review @polarathene just tell me so I can merge this.

LucidityCrash commented:
Thanks ... I was hoping to get time to work on it tomorrow, but it looks like you've done it for me :)

casperklein commented on target/scripts/startup/check-stack.sh:
```diff
@@ -60,3 +60,11 @@ function _check_log_level
+{
+  if [[ ${ENABLE_FETCHMAIL} -eq 1 ]] && [[ ${ENABLE_GETMAIL} -eq 1 ]]
+  then
+    _shutdown 'Enabling Fetchmail and Getmail at the same time is unsupported'
```
What's the reason for that? Why can't someone use both at the same time?

LucidityCrash replied:
Good question, I just kinda assumed people would use one or the other, I don't think there is an actual conflict that would stop it though.

LucidityCrash replied:
I've removed the exclusivity check.

georglauterbach commented:
> Thanks ... I was hoping to get time to work on it tomorrow, but it looks like you've done it for me :)

👍🏼 It'd be nice if you could add a check or two for the debug-getmail script, so we know what the expected output should look like :)

LucidityCrash commented:
> > Thanks ... I was hoping to get time to work on it tomorrow, but it looks like you've done it for me :)
> 
> 👍🏼 It'd be nice if you could add a check or two for the debug-getmail script, so we know what the expected output should look like :)

I've added a dummy config file for a getmail user and an output  check to see if the combined config file (user cfg and the general cfg) is correct.  I also added an output check for the dump command in debug-getmail.  Not 100% sure the way I've done it is correct ... there is leading white space on some of the returned lines.

LucidityCrash commented:
Tests fail with it saying output is not what is expected but looking at what is says expected is vs what it gets I can't see a difference.

georglauterbach commented:
It has always been a weakness of BATS to make it harder than it should be to match for multi-line output. I took a straight-forward, but ugly-looking approach. Hope this works now.

georglauterbach commented:
It seem the situation with `assert_line` has become a bit better. Could you investigate the current issue please @LucidityCrash: <https://github.com/docker-mailserver/docker-mailserver/actions/runs/4971930503/jobs/8896799292#step:6:24>. Seems unrelated to my changes.

georglauterbach commented on target/bin/debug-getmail:
```diff
@@ -0,0 +1,13 @@
+
+# shellcheck source=../scripts/helpers/log.sh
+source /usr/local/bin/helpers/log.sh
+# shellcheck source=../scripts/startup/setup-stack.sh
+source /usr/local/bin/setup-stack.sh
```
Just realized: the current error is related here. You'll probably not source this stack, but the corresponding setup script (<[target/scripts/startup/setup.d/getmail.sh](https://github.com/docker-mailserver/docker-mailserver/pull/2803/files#diff-9103a52f7cd4812e074e2240f8628053ba0d7ea91402c9f48b13c06f85fea06f)>) I guess.

LucidityCrash replied:
fixed that ...l my bad ... I'd done this on my local branch but overwrote it with the version you checked in :)

LucidityCrash commented:
Worked out how to run the tests myself on my own fork :)
Sorry still not working and I think I've identified why just don't know how to fix it in the context of the testing workflow.

Has mail_state.sh been run at this point?  as the output shows : 
# output : Error: Specified config/data dir "/var/lib/getmail" does not exist - create or specify alternate directory with --getmaildir option

this happens when the directory specified for the --getmaildir option in debug-getmail doesn't exist.
I could create the folder /var/lib/getmail in the debug-getmail script. But not sure if that is the right way to proceed ?

georglauterbach commented:
> Worked out how to run the tests myself on my own fork :) Sorry still not working and I think I've identified why just don't know how to fix it in the context of the testing workflow.
> 
> Has mail_state.sh been run at this point? as the output shows :
> # output : Error: Specified config/data dir "/var/lib/getmail" does not exist - create or specify alternate directory with --getmaildir option
>
> this happens when the directory specified for the --getmaildir option in debug-getmail doesn't exist. I could create the folder /var/lib/getmail in the debug-getmail script. But not sure if that is the right way to proceed ?

It seems to me that this is fine :) Because there is no directory mounted to `/var/mail-state` in a test container, the check `[[ ${ONE_DIR} -eq 1 ]] && [[ -d ${STATEDIR} ]]` in `mail_state.sh` does not succeed, and the directory is not linked. For the test it does not matter though whether changes are persisted 👍🏼

georglauterbach commented:
Probably the cleanest solution is to check in `debug-getmail` whether the directory exists in the first place. This is what I'd do :) If it does not exists, just use a directory that you know exists. In case `/var/lib/getmail` does not exists, we can reasonable assume that `/var/mail-state` is not mounted, and hence, there is no persistence in the container. This, in turn, warrants using something like `/tmp` for `--getmaildir` IMO.

LucidityCrash commented:
Sounds like a plan ... Will get on it as soon as I can ... probably be
tomorrow now.

On Sun, 14 May 2023, 16:50 Georg Lauterbach, ***@***.***>
wrote:

> Probably the cleanest solution is to check in debug-getmail whether the
> directory exists in the first place. This is what I'd do :) If it does not
> exists, just use a directory that you know exists. In case
> /var/lib/getmail does not exists, we can reasonable assume that
> /var/mail-state is not mounted, and hence, there is no persistence in the
> container. This, in turn, warrants using something like /tmp for
> --getmaildir IMO.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docker-mailserver/docker-mailserver/pull/2803#issuecomment-1546929299>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALWZJK3LQC4KLRNBBM66TYDXGD5ODANCNFSM6AAAAAAQY3HJXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

LucidityCrash commented:
It's still failing ... Will work on it again tomorrow

On Sun, 14 May 2023, 20:09 Georg Lauterbach, ***@***.***>
wrote:

> Looks good - maybe use mkdir -p instead of mkdir :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docker-mailserver/docker-mailserver/pull/2803#issuecomment-1546978480>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALWZJK53TKIDGYG3OZFVSALXGEUYHANCNFSM6AAAAAAQY3HJXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

LucidityCrash commented:
Actuall found a few minutes ... think I got it this time : 
``` 
ok 18 [Getmail] default configuration exists and is correct in 350ms
ok 19 [Getmail] debug-getmail works as expected in 993ms
```

and the tests in my fork are green :)

georglauterbach commented:
Looks good! I will have a look tomorrow for final review 🚀

georglauterbach reviewed (dismissed):
Looks good to me now :) Maybe @casperklein want to have another look at this as well, and I think there is one comment from him that should also be resolved. But overall, nice work 🚀

casperklein commented on docs/content/config/environment.md:
```diff
@@ -548,6 +548,20 @@ Note: activate this only if you are confident in your bayes database for identif
   **1** => `/etc/fetchmailrc` is split per poll entry. For every poll entry a separate fetchmail instance is started  to allow having multiple imap idle configurations defined.
 
 Note: The defaults of your fetchmailrc file need to be at the top of the file. Otherwise it won't be added correctly to all separate `fetchmail` instances.
+      Fetchmail and Getmail are mutually exclusive
+
+#### Getmail
+
+##### ENABLE_GETMAIL
+
+Enable or disable `getmail`. Note: Fetchmail ([`ENABLE_FETCHMAIL`](#enable_fetchmail)) and Getmail are mutually exclusive!
```
```suggestion
#### Getmail

##### ENABLE_GETMAIL

Enable or disable `getmail`.
```

casperklein commented on mailserver.env:
```diff
@@ -382,6 +382,16 @@ ENABLE_FETCHMAIL=0
 FETCHMAIL_POLL=300
 
+# Enable or disable `getmail`.
+# Note: Fetchmail (ENABLE_FETCHMAIL) and Getmail are mutually exclusive!
```
```suggestion
```

casperklein commented on docs/content/config/environment.md:
```diff
@@ -548,6 +548,18 @@ Note: activate this only if you are confident in your bayes database for identif
+
+##### GETMAIL_POLL
+
+The number of minutes for the interval. Min: 1; Max: 30; Default: 5.
```
```suggestion
- **5** => `getmail` The number of minutes for the interval. Min: 1; Max: 30; Default: 5.
```

casperklein commented on target/scripts/build/packages.sh:
```diff
@@ -201,6 +201,20 @@ function _install_fail2ban
 
+# Presently the getmail6 package is v6.14, which is too old.
+# v6.18 contains fixes for Google and Microsoft OAuth support.
+# using pip to install getmail.
```
```suggestion
# using pip to install getmail.
# TODO This can be removed when the base image is updated to Debian 12 (Bookworm)
```

casperklein commented on target/scripts/build/packages.sh:
```diff
@@ -201,6 +201,20 @@ function _install_fail2ban
+{
+  _log 'debug' 'Installing getmail6'
+  apt-get "${QUIET}" --no-install-recommends install python3-pip
+  pip3 install 'getmail6~=6.18.12'
```
```suggestion
  pip3 install --no-cache-dir 'getmail6~=6.18.12'
```

casperklein reviewed (changes requested):
Just some small change requests. Otherwise ready to merge.
